File: core/src/main/java/org/stepik/core/metrics/Metrics.java
package org.stepik.core.metrics;

import com.intellij.openapi.application.ApplicationInfo;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.jetbrains.tmp.learning.StepikProjectManager;
import com.jetbrains.tmp.learning.courseFormat.CourseNode;
import com.jetbrains.tmp.learning.courseFormat.StepNode;
import com.jetbrains.tmp.learning.stepik.StepikConnectorLogin;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.exceptions.StepikClientException;
import org.stepik.api.queries.metrics.StepikMetricsPostQuery;
import org.stepik.core.utils.PluginUtils;

import java.util.UUID;
import java.util.function.Consumer;

/**
 * @author meanmail
 */
public class Metrics {
    private static final Logger logger = Logger.getInstance(Metrics.class);
    private static final String session = UUID.randomUUID().toString();

    public static void downloadAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("download", project, stepNode, status);
    }

    private static void postMetrics(@NotNull Project project, @NotNull Consumer<StepikMetricsPostQuery> installer) {
        StepikMetricsPostQuery query = null;
        try {
            StepikApiClient stepikApiClient = StepikConnectorLogin.authAndGetStepikApiClient();

            query = stepikApiClient.metrics()
                    .post()
                    .name("ide_plugin")
                    .tags("name", "S_Union")
                    .tags("ide_name", ApplicationInfo.getInstance().getVersionName())
                    .data("ide_version", ApplicationInfo.getInstance().getBuild().toString())
                    .data("plugin_version", PluginUtils.getVersion())
                    .data("session", session);

            StepikProjectManager projectManager = StepikProjectManager.getInstance(project);

            if (projectManager != null) {
                query.data("project_id", projectManager.getUuid());
            }

            installer.accept(query);

            query.execute();
        } catch (StepikClientException e) {
            String message = String.format("Failed post metric: %s", query != null ? query.toString() : "null");
            logger.warn(message, e);
        }
    }

    public static void sendAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("send", project, stepNode, status);
    }

    public static void getStepStatusAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("get_step_status", project, stepNode, status);
    }

    private static void stepAction(
            @NotNull String actionName,
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        CourseNode courseNode = stepNode.getCourse();
        long courseId = courseNode != null ? courseNode.getId() : 0;

        Consumer<StepikMetricsPostQuery> installer = query -> query.tags("action", actionName)
                .data("course_id", courseId)
                .data("step_id", stepNode.getId())
                .tags("status", status)
                .tags("programming_language", stepNode.getCurrentLang().getName())
                .tags("step_type", stepNode.getData().getBlock().getName());

        postMetrics(project, installer);
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/metrics/Metrics.kt
package org.stepik.core.metrics

import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.components.ServiceManager.getService
import com.intellij.openapi.project.Project
import org.stepik.api.client.StepikApiClient
import org.stepik.api.objects.metrics.Metric
import org.stepik.core.PluginSettings
import org.stepik.core.SupportedLanguages
import org.stepik.core.auth.StepikAuthManager.authAndGetStepikApiClient
import org.stepik.core.auth.StepikAuthManager.isAuthenticated
import org.stepik.core.common.Loggable
import org.stepik.core.courseFormat.StepNode
import org.stepik.core.courseFormat.StudyNode
import org.stepik.core.getProjectManager
import org.stepik.core.metrics.MetricsStatus.SUCCESSFUL
import org.stepik.core.pluginMetricsName
import org.stepik.core.utils.currentProject
import org.stepik.core.utils.version
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

object Metrics : Loggable {
    private val session = UUID.randomUUID()
            .toString()
    private val executor = Executors.newScheduledThreadPool(1)
    private val pluginId = getService(PluginSettings::class.java).pluginId
    
    private fun postMetrics(
            project: Project,
            metric: Metric,
            status: MetricsStatus) {
        executor.schedule({
            val stepikApiClient: StepikApiClient = authAndGetStepikApiClient()
            if (!isAuthenticated) {
                return@schedule
            }
            
            val appInfo = ApplicationInfo.getInstance()
            
            val query = stepikApiClient.metrics()
                    .post()
                    .timestamp(System.currentTimeMillis() / 1000L)
                    .tags(metric.tags!!)
                    .data(metric.data!!)
                    .name("ide_plugin")
                    .tags("name", pluginMetricsName)
                    .tags("ide_name", appInfo.versionName)
                    .data("ide_version", appInfo.build.toString())
                    .data("plugin_version", version(pluginId))
                    .data("session", session)
                    .tags("status", status)
            
            val projectManager = getProjectManager(project)
            
            if (projectManager != null) {
                query.data("project_id", projectManager.getUuid())
                        .tags("project_programming_language",
                                (projectManager.defaultLang ?: SupportedLanguages.INVALID).langName)
                        .data("project_manager_version", projectManager.version)
                
                val projectRoot = projectManager.projectRoot
                
                if (projectRoot != null) {
                    val projectRootClass = projectRoot.javaClass
                    query.tags("project_root_class", projectRootClass.simpleName)
                            .data("project_root_id", projectRoot.id)
                    
                    query.data("course_id", projectRoot.getCourseId(stepikApiClient))
                }
            }
            
            query.executeAsync()
                    .exceptionally { e ->
                        val message = String.format("Failed post metric: %s", query.toString())
                        logger.warn(message, e)
                        null
                    }
        }, 500, TimeUnit.MILLISECONDS)
    }
    
    private fun postSimpleMetric(
            project: Project,
            action: String,
            status: MetricsStatus) {
        val metric = Metric().addTags("action", action)
        postMetrics(project, metric, status)
    }
    
    fun authenticate(status: MetricsStatus = SUCCESSFUL) {
        postSimpleMetric(currentProject, "authenticate", status)
    }
    
    fun createProject(project: Project, status: MetricsStatus = SUCCESSFUL) {
        postSimpleMetric(project, "create_project", status)
    }
    
    fun openProject(project: Project, status: MetricsStatus = SUCCESSFUL) {
        postSimpleMetric(project, "open_project", status)
    }
    
    private fun stepAction(
            actionName: String,
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        val metric = Metric().apply {
            addTags("action", actionName)
            addData("step_id", stepNode.id)
            addTags("step_programming_language", stepNode.currentLang.langName)
            addTags("step_type", stepNode.type.typeName)
        }
        postMetrics(project, metric, status)
    }
    
    fun sendAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("send", project, stepNode, status)
    }
    
    fun downloadAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("download", project, stepNode, status)
    }
    
    fun getStepStatusAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("get_step_status", project, stepNode, status)
    }
    
    fun insertAmbientCodeAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("insert_ambient_code", project, stepNode, status)
    }
    
    fun navigateAction(
            project: Project,
            studyNode: StudyNode,
            status: MetricsStatus = SUCCESSFUL) {
        if (studyNode is StepNode) {
            stepAction("navigate", project, studyNode, status)
        }
    }
    
    fun openInBrowserAction(
            project: Project,
            studyNode: StudyNode) {
        if (studyNode is StepNode) {
            stepAction("open_in_browser", project, studyNode, SUCCESSFUL)
        }
    }
    
    fun resetStepAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("reset_step", project, stepNode, status)
    }
    
    fun removeAmbientCodeAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("remove_ambient_code", project, stepNode, status)
    }
    
    fun switchLanguage(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("switch_language", project, stepNode, status)
    }
    
    fun testCodeAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus = SUCCESSFUL) {
        stepAction("test_code", project, stepNode, status)
    }
}
